### ISSUE TITLE
[Feature Request] Dynamic spark.databricks.delta.snapshotPartitions based on size of snapshot

### ISSUE THREAD

**Issue by santosh-d3vpl3x** [delta-io/delta#3351]
#### Which Delta project/connector is this regarding?
- [x] Spark
- [ ] Standalone
- [ ] Flink
- [ ] Kernel
- [ ] Other (fill in here)

### Overview

Currently, the `spark.databricks.delta.snapshotPartitions` value is static. The idea is to make this value dependent on the size of snapshot such that we can have well sized partitions.

### Motivation

Delta computes the snapshot to understand which parquet files to read and caches it to make it performant. The cached number of partitions depend upon `spark.databricks.delta.snapshotPartitions`. For bigger tables, the default value of 50 might be saner but for smaller tables, this usually results in partition sizes of few bytes. This does not play well with dynamic allocation. It is not recommended to kill an executor that has cached partition on it, by default spark sets decommission time to infinity for such executors. Many a times, this leads to idle executor staying alive just because it has few bytes of delta cache.

### Further details

A naïve approach: can we leverage AQE here?

### Willingness to contribute

The Delta Lake Community encourages new feature contributions. Would you or another member of your organization be willing to contribute an implementation of this feature?

- [ ] Yes. I can contribute this feature independently.
- [x] Yes. I would be willing to contribute this feature with guidance from the Delta Lake community.
- [ ] No. I cannot contribute this feature at this time.

**Comment by Kimahriman** [comment]
You can set the storage level for the snapshot cache: https://github.com/delta-io/delta/pull/1000, but this would also be a good feature. I've thought about doing this for a while but never got to it. You have all the file info for the files needed to load the snapshot I believe, so you can do some stats based on the total size of those files perhaps.

**Comment by santosh-d3vpl3x** [comment]
TIL! I couldn't find any reference to that config anywhere, thanks!

I wish delta config also used `version` to indicate since which version the config existed. Which version of delta has this change?

**Comment by Kimahriman** [comment]
Yeah a lot of the configs you just kinda need to know about or dig through the source code to find and see what they do. I specifically added that for the dynamic allocation issue you mentioned hah. I'm not sure why they don't have a version on the config, the only real way to tell is to look at what tag that commit made it into which it looks like you did, 1.2.0.

**Comment by Kimahriman** [comment]
Finally got around to trying to knock this out. Ignoring the cache issue, it would still be nice to dynamically size the snapshot partitions for small vs large tables

**Comment by foss-contributor** [comment]
@santosh-d3vpl3x, thank you for your contribution! This issue has been inactive for over 18 months and will be closed in one week. To keep it open, please leave a comment on this issue or contact us in the `#contributors` channel on [Delta Lake Slack](https://go.delta.io/slack). We appreciate your involvement in Delta Lake!
